File: modules/core-api/src/main/java/com/enonic/xp/query/expr/QueryExpr.java
package com.enonic.xp.query.expr;

import java.util.Arrays;
import java.util.List;

import com.google.common.annotations.Beta;
import com.google.common.base.Joiner;
import com.google.common.collect.ImmutableList;

@Beta
public final class QueryExpr
    implements Expression
{
    private final ConstraintExpr constraint;

    private final ImmutableList<OrderExpr> orderList;

    public QueryExpr( final ConstraintExpr constraint, final Iterable<OrderExpr> orderSet )
    {
        this.constraint = constraint;

        if ( orderSet != null )
        {
            this.orderList = ImmutableList.copyOf( orderSet );
        }
        else
        {
            this.orderList = ImmutableList.of();
        }
    }

    public QueryExpr( final Iterable<OrderExpr> orderSet )
    {
        this.constraint = null;

        if ( orderSet != null )
        {
            this.orderList = ImmutableList.copyOf( orderSet );
        }
        else
        {
            this.orderList = ImmutableList.of();
        }
    }


    public static QueryExpr from( final ConstraintExpr constraint, final OrderExpr... orderSet )
    {
        return new QueryExpr( constraint, Arrays.asList( orderSet ) );
    }

    public static QueryExpr from( final ConstraintExpr constraint, final Iterable<OrderExpr> orderSet )
    {
        return new QueryExpr( constraint, orderSet );
    }

    public ConstraintExpr getConstraint()
    {
        return this.constraint;
    }

    public List<OrderExpr> getOrderList()
    {
        return this.orderList;
    }

    @Override
    public String toString()
    {
        final StringBuilder str = new StringBuilder();

        if ( this.constraint != null )
        {
            str.append( this.constraint.toString() );
        }

        if ( !this.orderList.isEmpty() )
        {
            if ( str.length() > 0 )
            {
                str.append( " " );
            }

            str.append( "ORDER BY " );
            str.append( Joiner.on( ", " ).join( this.orderList ) );
        }

        return str.toString();
    }
}

File: modules/admin/admin-ui/src/main/resources/web/admin/common/js/query/expr/QueryExpr.ts
module api.query.expr {

    export class QueryExpr implements Expression {

        private constraint: ConstraintExpr;
        private orderList: OrderExpr[] = [];

        constructor(constraint: ConstraintExpr, orderList?: OrderExpr[]) {
            this.constraint = constraint;
            if (orderList) {
                this.orderList = orderList;
            }
        }

        getConstraint(): ConstraintExpr {
            return this.constraint;
        }

        getOrderList(): OrderExpr[] {
            return this.orderList;
        }

        toString() {
            let result: string = "";

            if (this.constraint != null) {
                result = result.concat(this.constraint.toString());
            }

            if (this.orderList.length > 0) {
                result = result.concat(" ORDER BY ");

                let sub = [];
                this.orderList.forEach((expr: OrderExpr) => {
                    sub.push(expr.toString());
                });
                result = result.concat(sub.join(", "));
            }

            return result;
        }
    }
}
